File: FunctionalArithmeticEvaluator/Parser.fs
module FSharpParser

open FSharpLexer

type NodeType =
    | Sum      = 1
    | Subtract = 2
    | Multiply = 3
    | Divide   = 4
    | Module   = 5

type Node =
    | OperationNode of OperationNode
    | ValueNode of float
and OperationNode = { Type: NodeType; Right: Node; Left: Node;  }

let private getNodeType value =
    match value with
    | "+" -> NodeType.Sum
    | "-" -> NodeType.Subtract
    | "*" -> NodeType.Multiply
    | "/" -> NodeType.Divide
    | _ ->   NodeType.Module

let private getPrecendence value = 
    match value with
    | "+" | "-" -> 1
    | _ -> 2

let parse tokens =
    let rec parseRec (tokens: list<Token>) precedence visited op =
        match tokens with
        | head :: tail ->
            match head.Type with
            | TokenType.Float -> parseRec tail precedence (head :: visited) op
            | _  ->
                let tokenPrecendence = getPrecendence head.Value
                if (tokenPrecendence > precedence) then
                    parseRec tail precedence (head :: visited) op                    
                else 
                    let nodeType = getNodeType head.Value
                    parseRec tail precedence [] (fun left ->
                        parseRec visited precedence [] (fun right ->
                            op(OperationNode({ Type = nodeType; Right = right; Left = left }))))
        | _ -> 
            match visited with    
            | head :: [] -> op (head.Value |> float |> ValueNode)        
            | _ -> parseRec visited 2 []  op   
    parseRec (List.rev tokens) 1 [] (fun x -> x)

File: FunctionalArithmeticEvaluator.CSharp/Parser.cs
using Microsoft.FSharp.Collections;
using static FunctionalArithmeticEvaluator.CSharp.CSharpLexer;

namespace FunctionalArithmeticEvaluator.CSharp
{
    public static class CSharpParser
    {
        public enum OperationType
        {
            Sum,
            Subtract,
            Multiply,
            Divide,
            Module
        }

        public abstract class Node { }
        public class FloatNode: Node
        {
            public FloatNode(float value)
            {
                Value = value;
            }

            public float Value { get; }
        }

        public class OperationNode : Node
        {
            public OperationNode(OperationType operationType, Node left, Node right)
            {
                OperationType = operationType;
                Left = left;
                Right = right;
            }

            public OperationType OperationType { get; }
            public Node Right { get; }
            public Node Left { get; }
        }

        private static FSharpList<Token> Empty => ListModule.Empty<Token>();

        private static OperationType ToType(this string op) => op switch
        {
            "+" => OperationType.Sum,
            "-" => OperationType.Subtract,
            "*" => OperationType.Multiply,
            "/" => OperationType.Divide,
            _ => OperationType.Module,
        };

        private static int GetPriority(this string op) => op == "+" || op == "-" ? 1 : 2;
        
        private static Node ParseRec(this FSharpList<Token> tokens, FSharpList<Token> visited, int priority)
        {
            Token head = tokens.Head;

            FSharpList<Token> Visit() => new FSharpList<Token>(head, visited);

            if (tokens.Length == 1)
            {
                if (visited.IsEmpty)
                {
                    return new FloatNode(float.Parse(head.Value));
                }

                return ListModule.Reverse(Visit()).ParseRec(Empty, 2);
            }

            if (head.Type == TokenType.Operation)
            {
                OperationType type = head.Value.ToType();
                if (head.Value.GetPriority() == priority)
                {
                    Node left = tokens.Tail.ParseRec(Empty, 1);
                    Node right = ListModule.Reverse(visited).ParseRec(Empty, 1);
                    return new OperationNode(type, left, right);
                }
            }

            return tokens.Tail.ParseRec(Visit(), priority);
        }

        public static Node Parse(this FSharpList<Token> tokens)
        {
            return ListModule.Reverse(tokens).ParseRec(Empty, 1);
        }
    }
}
